### PR TITLE
[Fix] TE RMSNorm Triton Kernel Optimization

### PR DESCRIPTION
**Pull request by AllenFarcas**
# Description

Optimize RMSNorm triton kernel for Qwen3 shapes.

Fixes # https://github.com/ROCm/frameworks-internal/issues/16614

## Type of change

- [x] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Pack multiple rows per program (ROWS_PER_PID, up to 16 for H≤128) instead of one program per row — slashes program-launch count for narrow-H shapes.
- Apply the same packing to the backward kernel (capped at 4 for H≤128), targeting the Qwen3 32768×128 QK-norm gradient pass.
- Eliminate redundant gamma reloads and launch overhead on narrow rows, the root cause of the 0.1× cliff on Qwen3 QK shapes.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by Micky774** [comment]
Based on the [shared benchmarks](https://amd.atlassian.net/wiki/spaces/~712020d71119f77f2e48d98822f3c648e08ab9/pages/1701381099/2026-01-01+Issue+16614) it looks like there's actually a regression in the triton kernels across some shapes e.g. 

`Qwen3-235B/hidden, 8192, 4096, bfloat16, fwd: 4211.3 GB/s --> 3573.2 GB/s` similarly for the `bwd` of the same config. Can you run a general benchmark comparing the updated triton implementation to the current one on `dev` so we can see what the tradeoffs are? Maybe we can optimize heuristics a bit.

File: transformer_engine/pytorch/triton_kernels/utils.py [hunk 1]
```diff
@@ -33,13 +33,21 @@ def get_inf_ln_sm_margin():
...
 
 
+# Per-device SM count cache. get_num_sms is called on every RMSNorm dispatch
+# (forward and backward), and get_device_properties is a driver round-trip;
+# the active device's SM count is constant, so cache it.
+_NUM_SMS_CACHE: "dict[int, int]" = {}
+
+
 def get_num_sms(sm_margin=None):
-    num_sms = torch.cuda.get_device_properties(
```
Comment (ipanfilo) [inline transformer_engine/pytorch/triton_kernels/utils.py:37]: Use get_sm_count from transformer_engine.pytorch.utils

**Review by ipanfilo — CHANGES_REQUESTED** [review]

**Comment by AllenFarcas** [comment]
@Micky774 I updated the benchmarks run and the results obtained [here](https://amd.atlassian.net/wiki/x/6wNpZQ) and solved all regressions.

**Review by ipanfilo — APPROVED** [review]
review n_rows to ROWS_PER_PID divisibility like suggested by bot

File: transformer_engine/pytorch/triton_kernels/norms_common.py [hunk 2]
```diff
@@ -157,6 +157,8 @@ def _rows_per_pid(rows, hidden, num_sms, cap_ladder):
             rpp = 4
         elif hidden <= 512 and rows % 2 == 0:
             rpp = 2
+    # The kernel packs rpp rows per program with no row-tail mask, so rpp must divide rows
+    assert rows % rpp == 0, f"rows_per_pid={rpp} does not divide rows={rows}"
```
Comment (Micky774) [inline transformer_engine/pytorch/triton_kernels/norms_common.py]: Why not add a mask to allow for arbitrary shapes?

**Review by Micky774 — APPROVED** [review]
nit: `row_mask` should just be `mask` since otherwise it makes it sound like what `row_valid` actually is.

Otherwise, once CI is green LGTM
